src/truncated/lognormal.jl

```diff
@@ -0,0 +1,52 @@
+# Moments of the truncated log-normal can be computed directly from the moment generating
+# function of the truncated normal:
+#   Let Y ~ LogNormal(μ, σ) truncated to (a, b). Then log(Y) ~ Normal(μ, σ) truncated
+#   to (log(a), log(b)), and E[Y^n] = E[(e^log(Y))^n] = E[e^(nlog(Y))] = mgf(log(Y), n).
+
+# Given `truncate(LogNormal(μ, σ), a, b)`, return `truncate(Normal(μ, σ), log(a), log(b))`
+function _truncnorm(d::Truncated{<:LogNormal})
+    μ, σ = params(d.untruncated)
+    T = partype(d)
+    a = d.lower === nothing ? nothing : log(T(minimum(d)))
+    b = d.upper === nothing ? nothing : log(T(maximum(d)))
+    return truncated(Normal(μ, σ), a, b)
+end
+
+mean(d::Truncated{<:LogNormal}) = mgf(_truncnorm(d), 1)
+
+function var(d::Truncated{<:LogNormal})
+    tn = _truncnorm(d)
+    # Ensure the variance doesn't end up negative, which can occur due to numerical issues
+    return max(mgf(tn, 2) - mgf(tn, 1)^2, 0)
```

Comment:
Repeated evaluation of `mgf` involves repeated calculations of the same (intermediate) quantities. But my fear is that optimizing this further will lead to less readable code...

Reply:
Yeah, I had the same thought and wasn't sure what to do about it so I just... didn't address it, haha